tests/federation/transport/test_knocking.py

```diff
@@ -0,0 +1,289 @@
+
+        # Extract the generated knock event json
+        knock_event = channel.json_body["event"]
+
+        # Check that the event has things we expect in it
+        self.assertEquals(knock_event["room_id"], room_id)
+        self.assertEquals(knock_event["sender"], fake_knocking_user_id)
+        self.assertEquals(knock_event["state_key"], fake_knocking_user_id)
+        self.assertEquals(knock_event["type"], EventTypes.Member)
+        self.assertEquals(
+            knock_event["content"]["membership"], KNOCK_UNSTABLE_IDENTIFIER
+        )
+
+        # Turn the event json dict into a proper event.
+        # We won't sign it properly, but that's OK as we stub out event auth in `prepare`
+        signed_knock_event = builder.create_local_event_from_event_dict(
+            self.clock,
+            self.hs.hostname,
+            self.hs.signing_key,
+            room_version=RoomVersions.MSC2403_DEV,
+            event_dict=knock_event,
+        )
+
+        # Convert our proper event back to json dict format
+        signed_knock_event_json = signed_knock_event.get_pdu_json(
+            self.clock.time_msec()
+        )
+
+        # Send the signed knock event into the room
+        request, channel = self.make_request(
+            "PUT",
+            "/_matrix/federation/unstable/%s/send_knock/%s/%s"
+            % (KNOCK_UNSTABLE_IDENTIFIER, room_id, signed_knock_event.event_id),
+            signed_knock_event_json,
+        )
+        self.assertEquals(200, channel.code, channel.result)
+
+        # Check that we got the stripped room state in return
+        room_state_events = channel.json_body["knock_state_events"]
+
+        # Validate the stripped room state events
+        check_knock_room_state_against_room_state(
+            self, room_state_events, expected_room_state
+        )
+
+
+def send_example_state_events_to_room(
+    testcase: HomeserverTestCase, hs: "HomeServer", room_id: str, sender: str,
+) -> OrderedDict:
+    """Adds some state a room. State events are those that should be sent to a knocking
+    user after they knock on the room, as well as some state that *shouldn't* be sent
+    to the knocking user.
+
+    Args:
+        testcase: The testcase that is currently active.
+        hs: The homeserver of the sender.
+        room_id: The ID of the room to send state into.
+        sender: The ID of the user to send state as. Must be in the room.
+
+    Returns:
+        The OrderedDict of event types and content that a user is expected to see
+        after knocking on a room.
+    """
+    # To set a canonical alias, we'll need to point an alias at the room first.
+    canonical_alias = "#fancy_alias:test"
+    testcase.get_success(
+        testcase.store.create_room_alias_association(
+            RoomAlias.from_string(canonical_alias), room_id, ["test"]
+        )
+    )
+
+    # Send some state that we *don't* expect to be given to knocking users
+    secret_state_event_type = "com.example.secret"
```

Comment:
Do we need the constant (`SECRET_STATE_EVENT_TYPE`) above?

Reply:
Ah, this isn't supposed to be here. We either have a constant or just initialise the same variable twice. I've gone with the constant route for the fix.